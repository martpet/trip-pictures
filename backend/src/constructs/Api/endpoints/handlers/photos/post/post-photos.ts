import {
  APIGatewayProxyHandlerV2,
  errorResponse,
  getIdTokenPayload,
  PhotosTableItem,
  photosTableOptions,
  PostPhotosRequest,
  PostPhotosResponse,
  RouteHeaders,
  StatusCodes,
} from 'lambda-layer';

import { ddbBatchGet, ddbBatchWrite } from '~/constructs/Api/utils';

export const handler: APIGatewayProxyHandlerV2<PostPhotosResponse> = async (event) => {
  const { authorization } = event.headers as RouteHeaders<'/settings'>;
  let requestData;

  if (!authorization) {
    return errorResponse('f31e4f94d3');
  }

  if (!event.body) {
    return errorResponse('24d59e00da', { statusCode: StatusCodes.BAD_REQUEST });
  }

  try {
    requestData = JSON.parse(event.body) as PostPhotosRequest;
  } catch (error) {
    return errorResponse('d45458a241', { statusCode: StatusCodes.BAD_REQUEST, error });
  }

  const existingItemsInDb = await ddbBatchGet<PhotosTableItem>({
    tableOptions: photosTableOptions,
    pkValues: requestData.map(({ fingerprint }) => fingerprint),
  });

  if (existingItemsInDb.length) {
    return errorResponse('9082bdd5f9', {
      statusCode: StatusCodes.FORBIDDEN,
      description: 'Cannot replace existing item',
    });
  }

  const { sub: userId } = await getIdTokenPayload(authorization);
  const createdAt = Date.now();

  return ddbBatchWrite({
    items: requestData.map((data) => ({ ...data, userId, createdAt })),
    tableOptions: photosTableOptions,
  });
};
