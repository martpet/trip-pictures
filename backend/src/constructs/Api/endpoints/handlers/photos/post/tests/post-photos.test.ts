import {
  ddbBatchGet,
  ddbBatchWrite,
  itCalls,
  itGetsIdToken,
  itHasJsonBody,
  itResolves,
  itResolvesWithError,
} from '~/constructs/Api/utils';

import { handler } from '../post-photos';

vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');
vi.mock('~/constructs/Api/utils/ddbHelpers/ddbBatchGet');
vi.mock('~/constructs/Api/utils/ddbHelpers/ddbBatchWrite');

beforeEach(() => {
  vi.mocked(ddbBatchGet).mockResolvedValue([]);
});

const args = [
  {
    headers: { authorization: 'dummyAuthorizationHeader' },
    body: JSON.stringify([{ fingerprint: 'dummyFingerprint1' }]),
  },
] as unknown as Parameters<typeof handler>;

describe('post-photos', () => {
  itHasJsonBody(handler, args);
  itGetsIdToken(handler, args);
  itCalls(ddbBatchGet, handler, args);
  itCalls(ddbBatchWrite, handler, args);
  itResolves(handler, args);

  describe('when "ddbBatchGet" has results', () => {
    beforeEach(() => {
      vi.mocked(ddbBatchGet).mockResolvedValue([{ fingerprint: 'dummyFingerPrint' }]);
    });
    itResolvesWithError(handler, args);
  });
});
