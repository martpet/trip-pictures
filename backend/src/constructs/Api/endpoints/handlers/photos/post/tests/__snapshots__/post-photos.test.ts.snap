// Vitest Snapshot v1

exports[`post-photos > calls "ddbBatchGet" with correct args 1`] = `
[
  [
    {
      "pkValues": [
        "dummyFingerprint1",
      ],
      "tableOptions": {
        "partitionKey": {
          "name": "fingerprint",
          "type": "S",
        },
        "stream": "NEW_IMAGE",
        "tableName": "TP-photos",
      },
    },
  ],
]
`;

exports[`post-photos > calls "ddbBatchWrite" with correct args 1`] = `
[
  [
    {
      "items": [
        {
          "createdAt": 1640995200000,
          "fingerprint": "dummyFingerprint1",
          "userId": "dummySub",
        },
      ],
      "tableOptions": {
        "partitionKey": {
          "name": "fingerprint",
          "type": "S",
        },
        "stream": "NEW_IMAGE",
        "tableName": "TP-photos",
      },
    },
  ],
]
`;

exports[`post-photos > calls "getIdTokenPayload" with correct args 1`] = `
[
  [
    "dummyAuthorizationHeader",
  ],
]
`;

exports[`post-photos > resolves with a correct value 1`] = `"dummyDdbBatchWriteResponse"`;

exports[`post-photos > when "authorization" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "f31e4f94d3",
  ],
]
`;

exports[`post-photos > when "authorization" header is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-photos > when "ddbBatchGet" has results > calls "errorResponse" with correct args 1`] = `
[
  [
    "9082bdd5f9",
    {
      "description": "Cannot replace existing item",
      "statusCode": 403,
    },
  ],
]
`;

exports[`post-photos > when "ddbBatchGet" has results > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-photos > when "event.body" is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "24d59e00da",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-photos > when "event.body" is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-photos > when "event.body" is not JSON > calls "errorResponse" with correct args 1`] = `
[
  [
    "d45458a241",
    {
      "error": [SyntaxError: Unexpected token o in JSON at position 1],
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-photos > when "event.body" is not JSON > resolves with a correct value 1`] = `"dummyErrorResponse"`;
